### ISSUE TITLE
How to configure key operation

### ISSUE THREAD

**Issue by kongmao45678** [grumpycoders/pcsx-redux#382]
Find a place where no buttons are configured, or can I only use the default configuration?

**Comment by nicolasnoble** [comment]
The input configuration dialog hasn't been completed yet: #318.

**Comment by wheremyfoodat** [comment]
> Find a place where no buttons are configured, or can I only use the default configuration?

@kongmao45678 Out of curiosity, are you using a keyboard or a controller with Redux?

**Comment by kongmao45678** [comment]
> > Find a place where no buttons are configured, or can I only use the default configuration?
> 
> @kongmao45678 Out of curiosity, are you using a keyboard or a controller with Redux?

Handle, keyboard occasionally

**Comment by lorn10** [comment]
@wheremyfoodat Out of curiosity, is there any xinput (aka Xbox 360 controller) support present in PCSX-Redux? I ask because my brand new 8BitDo Pro 2 controller doesn't work. It isn't detected, not in xinput nor dinput mode. :-( Well, this may be just a problem of me because I am most likely the only one on the planet which uses PCSX-Redux with Wine instead of building a native Linux binary. :-D Currently, only the Keyboard works for me.

Whatever, the **8BitDo Pro 2** controller is really an impressive piece of hardware. It's effectively "a state of the art" Bluetooth / USB 3 capable controller which works perfectly well especially for retro-console gaming. Ironically, although it looks 200% "PlayStation inspired", it is regarded currently as **the best** Nintendo switch gamepad. (It supports also an exclusive "Switch mode"). And of course, that controller works also absolutely perfectly (under Linux/Wine) in CXBX-R, the Xbox "emulator".

Long story short, this controller is the perfect device to play all the beloved PSX classics at the PC! I am waiting with excitement that #318 will be merged sometimes in the future. (The Wine devs should then have implemented any missing dinput stuff.) ;-)
![8BitDo_Pro_2](https://user-images.githubusercontent.com/9931771/120689781-c0027800-c4a4-11eb-8cc4-6509d2a2f805.png)
https://arstechnica.com/gaming/2021/04/8bitdo-pro-2-gamepad-review-a-50-bargain-for-cool-features-killer-performance/

**Comment by wheremyfoodat** [comment]
Hm, that's odd. I'd expect SDL to pick that controller up automagically without any issues.
Do you have the ability to build pcsx-redux on your machine to check if it's even getting detected and help search for possible issues? Don't know if any Redux contributor has one of those. @lorn10

**Comment by nicolasnoble** [comment]
It's supposed to be in the game controllers mapping list:

https://github.com/libsdl-org/SDL/blob/main/src/joystick/SDL_gamecontrollerdb.h#L44-L87

**Comment by lorn10** [comment]
The 8bitDo **Pro 2** controller model is brand new, released last month in Mai (or late April). So that maybe a reason why it isn't so widely common...

Regarding my problem, it is most likely Wine related. The "traditional" DInput gamepad stuff is currently not fully implemented in Wine. This stands in contrast to XInput which is in a better shape. Because of this I can use the controller (under Wine) in the Xbox emulator CXBX-R. I am not so familiar with this SDL and SDL2 stuff, according to the following replay (from a CXBX-R contributor) CXBX-R only uses DInput for keyboard and mouse while it uses XInput and SDL2 for the controllers:
https://github.com/grumpycoders/pcsx-redux/issues/474#issuecomment-829150128

But yes, it looks that I have really to learn my PCSX-redux compiling lesson (under Linux). Until now I have always avoided this.... In fact, only with a native Linux build I can cross-check under Kubuntu 20.04 LTS if the Pro 2 gamepad works. Will look when I find the time to experiment with this. Whatever, I am almost 100% sure that this controller will work, it offers really a wider range of communication protocols: **S** = Nintendo Switch, **A** = Android (special Android gamepad mode?), **D** = DirectInput and **X** = XInput
![8BitDo_Pro_2_Input_switch](https://user-images.githubusercontent.com/9931771/120842186-ab89b280-c56c-11eb-95db-725aa59a7d65.jpg)

**Comment by wheremyfoodat** [comment]
@lorn10 Do you have access to a windows machine perhaps? Running without a compatibility layer would help check if there's an issue with the 8BitDo Pro 2 (or SDL's Gamepad API) or if it's a wine problem. And as you know there's automatically provided Windows binaries so you wouldn't have to build anything.

Have you tried Redux with another controller before?

**Comment by nicolasnoble** [comment]
If it's brand new, there's a very real possibility it needs to get a mapping into SDL first. Right now, pcsx-redux only relies on SDL gamepad mapping, and nothing else. If it's not in the list, it doesn't exist for Redux.

**Comment by lorn10** [comment]
Yes, but I was unsuccessful because the available Windows 7 iMac 2006 is "a potato" in terms it has only a Direct X 9 GPU so no OpenGL 3.x. But I will try this at one other newer iMac with DX 10 class GPU and Windows 7 (native). All other available computers are already on Kubuntu 20.04 LTS.

Here follows a little `dmesg` output regarding this controller in **Bluetooth** (BT HID Gamepad), **DInput**, and **XInput** mode:
```
[83816.240522] input: 8BitDo Pro 2 as /devices/pci0000:00/0000:00:1a.7/usb1/1-1/1-1.1/1-1.1.1/1-1.1.1:1.0/bluetooth/hci0/hci0:12/0005:045E:02E0.0006/input/input18
[83816.240687] microsoft 0005:045E:02E0.0006: input,hidraw4: BLUETOOTH HID v9.03 Gamepad [8BitDo Pro 2] on 7c:d1:c3:9e:3f:2a
[83820.964840] usb 1-1.3: new full-speed USB device number 13 using ehci-pci
[83821.076146] usb 1-1.3: New USB device found, idVendor=2dc8, idProduct=6003, bcdDevice= 0.01
[83821.076150] usb 1-1.3: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[83821.076152] usb 1-1.3: Product: 8BitDo Pro 2
[83821.076153] usb 1-1.3: Manufacturer: 8BitDo
[83821.076155] usb 1-1.3: SerialNumber: 0000-0000
[83821.078530] input: 8BitDo 8BitDo Pro 2 as /devices/pci0000:00/0000:00:1a.7/usb1/1-1/1-1.3/1-1.3:1.0/0003:2DC8:6003.0007/input/input19
[83821.137175] hid-generic 0003:2DC8:6003.0007: input,hidraw4: USB HID v1.11 Gamepad [8BitDo 8BitDo Pro 2] on usb-0000:00:1a.7-1.3/input0
[83840.579915] usb 1-1.3: USB disconnect, device number 13
[83847.537113] usb 1-1.3: new full-speed USB device number 14 using ehci-pci
[83847.650672] usb 1-1.3: New USB device found, idVendor=045e, idProduct=028e, bcdDevice= 1.14
[83847.650676] usb 1-1.3: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[83847.650678] usb 1-1.3: Product: Controller
[83847.650680] usb 1-1.3: Manufacturer: 8BitDo
[83847.650681] usb 1-1.3: SerialNumber: 157F8F8
[83847.651072] input: Microsoft X-Box 360 pad as /devices/pci0000:00/0000:00:1a.7/usb1/1-1/1-1.3/1-1.3:1.0/input/input20
```

**Comment by nicolasnoble** [comment]
If it's identifying itself as an X-Box 360 pad, SDL would be able to use it straight.

I'll try and finish the PR from @wheremyfoodat to handle these things better.

**Comment by wheremyfoodat** [comment]
Yeah, most controls masquerading as Xbox controllers should be fine, though I've seen some less-known ones have issues in other emulators as well.
Hopefully this won't be too hard to fix.

**Comment by nicolasnoble** [comment]
For the original issue of reconfiguring key operations, #393 has now been merged in, thus fixing this one here.

**Comment by lorn10** [comment]
Great work @nicolasnoble :+1:  Just for reference, my **8bitDo Pro 2** controller issue is **also fixed**. 

So the gamepad works perfectly well (in XInput mode). This is true for my Windows 7 test system, but not for Wine were the latest PCSX-Redux builds are again broken with error message `kernelbase:AppPolicyGetProcessTerminationMethod FFFFFFFFFFFFFFFA, 000000000011FD70` Whatever, maybe this will work again in a feature Wine build...

**Comment by nicolasnoble** [comment]
Have you tried the appimage Linux binary distribution?

**Comment by lorn10** [comment]
Thanks for the hint. I have tested now the latest build **e277597b** and it works perfectly stable! Awesome. :+1:  (Yes, it works better than any previously in Wine tested Windows binary.) My 8bitDo Pro 2 is detected and configured (in XInput mode) right "out of the box".

![PCSX-redux(e277597b)nativeLINUX](https://user-images.githubusercontent.com/9931771/122651979-7cfb0280-d13c-11eb-9f52-cefb56100f01.png)

It's really really great that there now exist also native Linux binaries. :+1: This makes it much easier for the "normal-user folks". You just have to download PCSX-redux to a directory and start it. A little HowTo, for new Linux users:

1.) Download the appimage Linux binary to a directory of your choice and extract it there. KDE users may have to start it from terminal, so switch to the corresponding directory:
`cd $HOME/Downloads/PCSX-redux`

2.) Start it by putting a **./** before the filename, enjoy!
`./PCSX-Redux-HEAD-x86_64.AppImage`

Last thing, regarding the Linux appimage binaries, the latest one should be on top. Currently this isn't the case, so I downloaded an older one, build 03933413.
